File: C#/CommonExtensions/MatrixExtensions.cs
using MathNet.Numerics.LinearAlgebra;

namespace ExactDerivativesAndInitApproach
{
    public static class Extensions
    {
        public static Vector<double> Multiply(this Vector<double> first, Vector<double> second, double constant = 1.0)
        {
            var result = Vector<double>.Build.Dense(first.Count);

            for (int i = 0; i < first.Count; ++i)
            {
                result[i] = first[i] * second[i] * constant;
            }
            return result;
        }

        public static Matrix<double> Multiply(this Matrix<double> first, Matrix<double> second)
        {
            int rows = first.RowCount;
            int cols = first.ColumnCount;
            var result = Matrix<double>.Build.Dense(rows, cols);

            for (int i = 0; i < rows; ++i)
            {
                for (int j = 0; j < cols; ++j)
                {
                    for (int k = 0; k < cols; ++k)
                    {
                        result[i, j] += first[i, k] * second[k, j];
                    }
                }
            }
            return result;
        }

        public static Matrix<double> Subtract(this Matrix<double> first, Matrix<double> second)
        {
            int rows = first.RowCount;
            int cols = first.ColumnCount;
            var result = Matrix<double>.Build.Dense(rows, cols);

            for (int i = 0; i < rows; ++i)
            {
                for (int j = 0; j < cols; ++j)
                {
                    result[i, j] = first[i, j] - second[i, j];
                }
            }

            return result;
        }

        public static Vector<double> Sum(this Vector<double> first, Vector<double> second)
        {
            var result = Vector<double>.Build.Dense(first.Count);

            for (int i = 0; i < first.Count; ++i)
            {
                result[i] = first[i] + second[i];
            }

            return result;
        }

        public static Matrix<double> SetDiagonalElements(this Matrix<double> matrix, double value)
        {
            var n = matrix.RowCount;

            for (int i = 0; i < n; ++i)
            {
                matrix[i, i] = value;
            }

            return matrix;
        }

        public static Vector<double> GetRow(this Matrix<double> matrix, int row)
        {
            int cols = matrix.RowCount;
            var res = Vector<double>.Build.Dense(cols);

            for (int i = 0; i < cols; ++i)
            {
                res[i] = matrix[row, i];
            }

            return res;
        }

        public static Vector<double> GetColumn(this Matrix<double> matrix, int column)
        {
            int rows = matrix.RowCount;
            var res = Vector<double>.Build.Dense(rows);

            for (int i = 0; i < rows; ++i)
            {
                res[i] = matrix[i, column];
            }

            return res;
        }

        public static double SumFirstElements(this Vector<double> vector, int size)
        {
            double res = 0;
            for (int i = 0; i < size; ++i)
            {
                res += vector[i];
            }

            return res;
        }
    }
}

File: Common.Extensions/MatrixExtensions.fs
namespace Common.Extensions

open MathNet.Numerics

open System.Runtime.CompilerServices

module MatrixExtensions =
    let MultiplyVectors(first: LinearAlgebra.Vector<double>, second: LinearAlgebra.Vector<double>) =
        let result = LinearAlgebra.Vector<double>.Build.Dense(first.Count)

        for i = 0 to first.Count - 1 do
            result.Item(i) <- first.[i] * second.[i]

        result

    let DivideVectors(first: LinearAlgebra.Vector<double>, second: LinearAlgebra.Vector<double>) =
        let result = LinearAlgebra.Vector<double>.Build.Dense(first.Count)

        for i = 0 to first.Count - 1 do
            result.Item(i) <- first.[i] / second.[i]

        result

    let MultiplyVectorsTansponential(first: LinearAlgebra.Vector<double>, second: LinearAlgebra.Vector<double>) =
        let mutable result = 0.0

        for i = 0 to first.Count - 1 do
            result <- result + (first.[i] * second.[i])

        result

    let MultiplyMatrix(first: LinearAlgebra.Matrix<double>, second: LinearAlgebra.Matrix<double>) =
        let rows = first.RowCount
        let cols = first.ColumnCount
        let result = LinearAlgebra.Matrix<double>.Build.Dense(rows, cols)
        
        for i = 0 to rows - 1 do
            for j = 0 to cols - 1 do
                for k = 0 to cols - 1 do
                    result.Item(i, j) <- result.[i, j] + first.[i, k] * second.[k, j]

        result

    let Subtract(first: LinearAlgebra.Matrix<double>, second: LinearAlgebra.Matrix<double>) =
        let rows = first.RowCount
        let cols = first.ColumnCount
        let result = LinearAlgebra.Matrix<double>.Build.Dense(rows, cols)
        
        for i = 0 to rows - 1 do
            for j = 0 to cols - 1 do
                result.Item(i, j) <- first.[i, j] - second.[i, j]
        
        result

    let SubtractVectors(first: LinearAlgebra.Vector<double>, second: LinearAlgebra.Vector<double>) =
        let rows = first.Count
        let result = LinearAlgebra.Vector<double>.Build.Dense(rows)
    
        for i = 0 to rows - 1 do
            result.Item(i) <- first.[i] - second.[i]
    
        result

    let SubtractValue(first: LinearAlgebra.Vector<double>, value: double) =
        let rows = first.Count
        let result = LinearAlgebra.Vector<double>.Build.Dense(rows)

        for i = 0 to rows - 1 do
            result.Item(i) <- first.[i] - value

        result

    let InverseVector(vector: LinearAlgebra.Vector<double>) =
        let rows = vector.Count
        let result = LinearAlgebra.Vector<double>.Build.Dense(rows)

        for i = 0 to rows - 1 do
            result.Item(i) <- 1.0 / vector.[i]

        result

    let Sum(first: LinearAlgebra.Vector<double>, second: LinearAlgebra.Vector<double>) =
        let result = LinearAlgebra.Vector<double>.Build.Dense(first.Count)
        
        for i = 0 to first.Count - 1 do
            result.[i] <- first.[i] + second.[i]
        
        result

    let SetDiagonalElements(matrix: LinearAlgebra.Matrix<double>, value: double) =
        let n = matrix.RowCount

        for i = 0 to n - 1 do
            matrix.[i, i] <- value

        matrix

    let GetRow(matrix: LinearAlgebra.Matrix<double>, row: int) =
        let cols = matrix.RowCount
        let res = LinearAlgebra.Vector<double>.Build.Dense(cols)

        for i = 0 to cols - 1 do
            res.[i] <- matrix.[row, i]

        res

    let GetColumn(matrix: LinearAlgebra.Matrix<double>, column: int) =
        let rows = matrix.RowCount
        let res = LinearAlgebra.Vector<double>.Build.Dense(rows)

        for i = 0 to rows - 1 do
            res.[i] <- matrix.[i, column]

        res

    let SumFirstElements(vector: LinearAlgebra.Vector<double>, size: int) =
        let mutable res = 0.0;

        for i = 0 to size - 1 do
            res <- res + vector.[i]

        res

[<Extension>]
type MatrixExtensionsss () =
    [<Extension>]
    static member SumFirstElements(ref: Ref<LinearAlgebra.Vector<double>>, size: int) =
        let mutable res = 0.0;
    
        for i = 0 to size do
            res <- res + ref.Value.[i]

        res

    [<Extension>]
    static member MultiplyVectors(first: Ref<LinearAlgebra.Vector<double>>, second: LinearAlgebra.Vector<double>) =
        let result = LinearAlgebra.Vector<double>.Build.Dense(first.Value.Count)
        
        for i = 0 to first.Value.Count do
            result.Item(i) <- first.Value.[i] * second.[i]
        
        result
    
    [<Extension>]
    static member MultiplyMatrix(first: Ref<LinearAlgebra.Matrix<double>>, second: LinearAlgebra.Matrix<double>) =
        let rows = first.Value.RowCount
        let cols = first.Value.ColumnCount
        let result = LinearAlgebra.Matrix<double>.Build.Dense(rows, cols)

        for i = 0 to rows do
            for j = 0 to cols do
                for k = 0 to cols do
                    result.Item(i, j) <- result.[i, j] + first.Value.[i, k] * second.[k, j]

        result

    [<Extension>]
    static member Subtract(first: Ref<LinearAlgebra.Matrix<double>>, second: LinearAlgebra.Matrix<double>) =
        let rows = first.Value.RowCount
        let cols = first.Value.ColumnCount
        let result = LinearAlgebra.Matrix<double>.Build.Dense(rows, cols)

        for i = 0 to rows do
            for j = 0 to cols do
                result.Item(i, j) <- first.Value.[i, j] - second.[i, j]

        result

    [<Extension>]
    static member Sum(first: Ref<LinearAlgebra.Vector<double>>, second: LinearAlgebra.Vector<double>) =
        let result = LinearAlgebra.Vector<double>.Build.Dense(first.Value.Count)

        for i = 0 to first.Value.Count do
            result.[i] <- first.Value.[i] + second.[i]

        result

    [<Extension>]
    static member SetDiagonalElements(matrix: Ref<LinearAlgebra.Matrix<double>>, value: double) =
        let n = matrix.Value.RowCount
        
        for i = 0 to n do
            matrix.Value.[i, i] <- value
        
        matrix.Value

    [<Extension>]
    static member GetRow(matrix: Ref<LinearAlgebra.Matrix<double>>, row: int) =
        let cols = matrix.Value.RowCount
        let res = LinearAlgebra.Vector<double>.Build.Dense(cols)
        
        for i = 0 to cols do
            res.[i] <- matrix.Value.[row, i]
        
        res